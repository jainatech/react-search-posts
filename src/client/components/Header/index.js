import React from 'react'
import PropTypes from 'prop-types'
import { Navbar, Form, FormControl } from 'react-bootstrap';

function Header(props) {
    const searchRef = React.createRef();

    const doSearch = () => {
        if (searchRef && searchRef.current && searchRef.current.value) {
            props.filterData(searchRef.current.value);
        }
    }

    return (
        <>
            <Navbar expand="lg" variant="light" bg="light" className="bg-light justify-content-between">
                <Navbar.Brand>{props.text}</Navbar.Brand>
                {props.search ?
                    <Form inline>
                        <FormControl ref={searchRef} type="text" placeholder="Search by title" className=" mr-sm-2" onChange={doSearch} />
                    </Form> : null}

            </Navbar>
        </>
    )
}

Header.propTypes = {
    text: PropTypes.string,
    filterData: PropTypes.func
}

export default Header

