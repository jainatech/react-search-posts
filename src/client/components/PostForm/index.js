import React from "react";
import { Formik, Field, Form, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import PropTypes from 'prop-types';
import { Button } from 'react-bootstrap';

const PostForm = (props) => {
  const {editItem,editPost} = props;
  return (
    <Formik
      enableReinitialize
      initialValues={{
          title: editItem.title,
          body: editItem.body                   
      }}
      validationSchema={Yup.object().shape({
          title: Yup.string()
              .required('Title is required'),
          body: Yup.string()
              .required('Body is required'),
      })}
      onSubmit={fields => {
          alert('SUCCESS!! :-)\n\n' + JSON.stringify(fields, null, 4));
          const updatedData = { ...editItem, title: fields.title, body: fields.body }
          editPost(updatedData);
      }}
      >
      {({ errors, status, touched }) => (
          <Form>
            <div className="form-row">
                <div className="form-group col-11">
                    <label htmlFor="title">Title : </label>
                    <Field name="title" type="text" className={'form-control' + (errors.title && touched.title ? ' is-invalid' : '')} />
                    <ErrorMessage name="title" component="div" className="invalid-feedback" />
                </div>
                <div className="form-group col-11">
                    <label htmlFor="body">Body : </label>
                    <Field name="body" type="text" className={'form-control' + (errors.body && touched.body ? ' is-invalid' : '')} />
                    <ErrorMessage name="body" component="div" className="invalid-feedback" />
                </div>
            </div>
            <div className="form-group">
                <Button type="submit" className="btn btn-primary mr-2">Save</Button>
            </div>
          </Form>
      )}
    </Formik>
  );
}

PostForm.propTypes = {
  editItem: PropTypes.object,
  editPost: PropTypes.func
};

export default PostForm;