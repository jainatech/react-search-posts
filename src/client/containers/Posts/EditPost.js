import React from 'react'
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types'
import { Container, Col, Row } from 'react-bootstrap';
import { getPost, editPost } from "../../store/actions/posts";
import ENUMS from '../../utils/ENUMS';
import PostForm from '../../components/PostForm';
import Header from '../../components/Header';

const EditPost = (props) => {
    const id = props && props.match && props.match.params && props.match.params.id;
    const { posts, editPost, getPost, history } = props;
    const getPostData = (editItemId) => {
        if (posts.length === 0) {
            getPost();
        }
        const editItem = posts.filter(item => Number(item.id) === Number(editItemId));
        if (editItem.length > 0) {
            return editItem[0];
        }
        return {};
    }

    const doEdit = (updatedData) => {
        editPost(updatedData);
        history.push({
            pathname: ENUMS.ROUTES.HOME
        });
    }

    let editItem = getPostData(id);

    return (
        <>
        <Header text="Edit post details"  />

        <Container className="p-4">
            {/* <Row>
                <Col>
                    <h5 className="text-center">
                        Edit post details:
            </h5>
                </Col>
            </Row> */}
            <Row>
                <Col>
                    <PostForm editItem={editItem} editPost={doEdit}></PostForm>
                </Col>
            </Row>
        </Container>
        </>
        
    )
}

const mapDispatchToProps = dispatch => bindActionCreators({
    getPost: getPost,
    editPost: editPost
}, dispatch);

const mapStateToProps = (state) => ({
    posts: state.postsReducer.posts_data,
});

EditPost.propTypes = {
    editPost: PropTypes.func,
    getPost: PropTypes.func,
    posts: PropTypes.array
};

export default connect(mapStateToProps, mapDispatchToProps)(EditPost);
