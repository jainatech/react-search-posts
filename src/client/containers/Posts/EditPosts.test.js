import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import EditPost from './EditPost';
import PostForm from '../../components/PostForm';
import Header from '../../components/Header';

const mockStore = configureStore();
const userStore = mockStore({
  postsReducer: {
    posts_data: [{ "userId": 1, "id": 1, "title": "title1", "body": "body1" }, { "userId": 1, "id": 2, "title": "title2", "body": "body2" }],
    loading: true
  }
});
const createTestProps = (props) => ({
  match : {
    params : {
      id: 1
    }
  },
  ...props
})

describe('EditPost module case--->', () => {
  it('EditPost match snapshot', () => {
    let props = createTestProps({
      posts: [{ "userId": 1, "id": 1, "title": "title1", "body": "body1" }, { "userId": 1, "id": 2, "title": "title2", "body": "body2" }],
      loading: true
    })
    let wrapper = mount(<Provider store={userStore}><EditPost {...props}></EditPost></Provider>);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('EditPost run without crashing', () => {
    const component = shallow(<Provider store={userStore}><EditPost /></Provider>);
    expect(component.find(EditPost)).toBeDefined();
  });

  it('EditPost run without crashing', () => {
    const component = shallow(<Provider store={userStore}><EditPost ><Header/><PostForm/></EditPost></Provider>);
    expect(component.find(EditPost)).toBeDefined();
  });

});