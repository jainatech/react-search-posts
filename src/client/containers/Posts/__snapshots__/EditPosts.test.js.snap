// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPost module case---> EditPost match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(EditPost)
    loading={true}
    match={
      Object {
        "params": Object {
          "id": 1,
        },
      }
    }
    posts={
      Array [
        Object {
          "body": "body1",
          "id": 1,
          "title": "title1",
          "userId": 1,
        },
        Object {
          "body": "body2",
          "id": 2,
          "title": "title2",
          "userId": 1,
        },
      ]
    }
  >
    <EditPost
      editPost={[Function]}
      getPost={[Function]}
      loading={true}
      match={
        Object {
          "params": Object {
            "id": 1,
          },
        }
      }
      posts={
        Array [
          Object {
            "body": "body1",
            "id": 1,
            "title": "title1",
            "userId": 1,
          },
          Object {
            "body": "body2",
            "id": 2,
            "title": "title2",
            "userId": 1,
          },
        ]
      }
    >
      <Header
        text="Edit post details"
      >
        <Navbar
          bg="light"
          className="bg-light justify-content-between"
          collapseOnSelect={false}
          expand="lg"
          variant="light"
        >
          <nav
            className="bg-light justify-content-between navbar navbar-expand-lg navbar-light bg-light"
          >
            <NavbarBrand>
              <span
                className="navbar-brand"
              >
                Edit post details
              </span>
            </NavbarBrand>
          </nav>
        </Navbar>
      </Header>
      <Container
        className="p-4"
        fluid={false}
      >
        <div
          className="p-4 container"
        >
          <Row
            noGutters={false}
          >
            <div
              className="row"
            >
              <Col>
                <div
                  className="col"
                >
                  <PostForm
                    editItem={
                      Object {
                        "body": "body1",
                        "id": 1,
                        "title": "title1",
                        "userId": 1,
                      }
                    }
                    editPost={[Function]}
                  >
                    <Formik
                      enableReinitialize={true}
                      initialValues={
                        Object {
                          "body": "body1",
                          "title": "title1",
                        }
                      }
                      onSubmit={[Function]}
                      validationSchema={
                        ObjectSchema {
                          "_blacklist": RefSet {
                            "list": Set {},
                            "refs": Map {},
                          },
                          "_conditions": Array [],
                          "_defaultDefault": [Function],
                          "_deps": Array [],
                          "_excludedEdges": Array [],
                          "_exclusive": Object {},
                          "_mutate": undefined,
                          "_nodes": Array [
                            "body",
                            "title",
                          ],
                          "_options": Object {
                            "abortEarly": true,
                            "recursive": true,
                          },
                          "_type": "object",
                          "_typeError": [Function],
                          "_whitelist": RefSet {
                            "list": Set {},
                            "refs": Map {},
                          },
                          "fields": Object {
                            "body": StringSchema {
                              "_blacklist": RefSet {
                                "list": Set {},
                                "refs": Map {},
                              },
                              "_conditions": Array [],
                              "_deps": Array [],
                              "_exclusive": Object {
                                "required": true,
                              },
                              "_mutate": undefined,
                              "_options": Object {
                                "abortEarly": true,
                                "recursive": true,
                              },
                              "_type": "string",
                              "_typeError": [Function],
                              "_whitelist": RefSet {
                                "list": Set {},
                                "refs": Map {},
                              },
                              "tests": Array [
                                [Function],
                              ],
                              "transforms": Array [
                                [Function],
                              ],
                              "type": "string",
                            },
                            "title": StringSchema {
                              "_blacklist": RefSet {
                                "list": Set {},
                                "refs": Map {},
                              },
                              "_conditions": Array [],
                              "_deps": Array [],
                              "_exclusive": Object {
                                "required": true,
                              },
                              "_mutate": undefined,
                              "_options": Object {
                                "abortEarly": true,
                                "recursive": true,
                              },
                              "_type": "string",
                              "_typeError": [Function],
                              "_whitelist": RefSet {
                                "list": Set {},
                                "refs": Map {},
                              },
                              "tests": Array [
                                [Function],
                              ],
                              "transforms": Array [
                                [Function],
                              ],
                              "type": "string",
                            },
                          },
                          "tests": Array [],
                          "transforms": Array [
                            [Function],
                          ],
                          "type": "object",
                        }
                      }
                    >
                      <Form>
                        <form
                          action="#"
                          onReset={[Function]}
                          onSubmit={[Function]}
                        >
                          <div
                            className="form-row"
                          >
                            <div
                              className="form-group col-11"
                            >
                              <label
                                htmlFor="title"
                              >
                                Title : 
                              </label>
                              <Field
                                className="form-control"
                                name="title"
                                type="text"
                              >
                                <input
                                  className="form-control"
                                  name="title"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  type="text"
                                  value="title1"
                                />
                              </Field>
                              <FormikConnect(ErrorMessageImpl)
                                className="invalid-feedback"
                                component="div"
                                name="title"
                              >
                                <ErrorMessageImpl
                                  className="invalid-feedback"
                                  component="div"
                                  formik={
                                    Object {
                                      "dirty": false,
                                      "errors": Object {},
                                      "getFieldHelpers": [Function],
                                      "getFieldMeta": [Function],
                                      "getFieldProps": [Function],
                                      "handleBlur": [Function],
                                      "handleChange": [Function],
                                      "handleReset": [Function],
                                      "handleSubmit": [Function],
                                      "initialErrors": Object {},
                                      "initialStatus": undefined,
                                      "initialTouched": Object {},
                                      "initialValues": Object {
                                        "body": "body1",
                                        "title": "title1",
                                      },
                                      "isSubmitting": false,
                                      "isValid": true,
                                      "isValidating": false,
                                      "registerField": [Function],
                                      "resetForm": [Function],
                                      "setErrors": [Function],
                                      "setFieldError": [Function],
                                      "setFieldTouched": [Function],
                                      "setFieldValue": [Function],
                                      "setFormikState": [Function],
                                      "setStatus": [Function],
                                      "setSubmitting": [Function],
                                      "setTouched": [Function],
                                      "setValues": [Function],
                                      "status": undefined,
                                      "submitCount": 0,
                                      "submitForm": [Function],
                                      "touched": Object {},
                                      "unregisterField": [Function],
                                      "validateField": [Function],
                                      "validateForm": [Function],
                                      "validateOnBlur": true,
                                      "validateOnChange": true,
                                      "validateOnMount": false,
                                      "values": Object {
                                        "body": "body1",
                                        "title": "title1",
                                      },
                                    }
                                  }
                                  name="title"
                                />
                              </FormikConnect(ErrorMessageImpl)>
                            </div>
                            <div
                              className="form-group col-11"
                            >
                              <label
                                htmlFor="body"
                              >
                                Body : 
                              </label>
                              <Field
                                className="form-control"
                                name="body"
                                type="text"
                              >
                                <input
                                  className="form-control"
                                  name="body"
                                  onBlur={[Function]}
                                  onChange={[Function]}
                                  type="text"
                                  value="body1"
                                />
                              </Field>
                              <FormikConnect(ErrorMessageImpl)
                                className="invalid-feedback"
                                component="div"
                                name="body"
                              >
                                <ErrorMessageImpl
                                  className="invalid-feedback"
                                  component="div"
                                  formik={
                                    Object {
                                      "dirty": false,
                                      "errors": Object {},
                                      "getFieldHelpers": [Function],
                                      "getFieldMeta": [Function],
                                      "getFieldProps": [Function],
                                      "handleBlur": [Function],
                                      "handleChange": [Function],
                                      "handleReset": [Function],
                                      "handleSubmit": [Function],
                                      "initialErrors": Object {},
                                      "initialStatus": undefined,
                                      "initialTouched": Object {},
                                      "initialValues": Object {
                                        "body": "body1",
                                        "title": "title1",
                                      },
                                      "isSubmitting": false,
                                      "isValid": true,
                                      "isValidating": false,
                                      "registerField": [Function],
                                      "resetForm": [Function],
                                      "setErrors": [Function],
                                      "setFieldError": [Function],
                                      "setFieldTouched": [Function],
                                      "setFieldValue": [Function],
                                      "setFormikState": [Function],
                                      "setStatus": [Function],
                                      "setSubmitting": [Function],
                                      "setTouched": [Function],
                                      "setValues": [Function],
                                      "status": undefined,
                                      "submitCount": 0,
                                      "submitForm": [Function],
                                      "touched": Object {},
                                      "unregisterField": [Function],
                                      "validateField": [Function],
                                      "validateForm": [Function],
                                      "validateOnBlur": true,
                                      "validateOnChange": true,
                                      "validateOnMount": false,
                                      "values": Object {
                                        "body": "body1",
                                        "title": "title1",
                                      },
                                    }
                                  }
                                  name="body"
                                />
                              </FormikConnect(ErrorMessageImpl)>
                            </div>
                          </div>
                          <div
                            className="form-group"
                          >
                            <Button
                              active={false}
                              className="btn btn-primary mr-2"
                              disabled={false}
                              type="submit"
                              variant="primary"
                            >
                              <button
                                className="btn btn-primary mr-2 btn btn-primary"
                                disabled={false}
                                type="submit"
                              >
                                Save
                              </button>
                            </Button>
                          </div>
                        </form>
                      </Form>
                    </Formik>
                  </PostForm>
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
    </EditPost>
  </Connect(EditPost)>
</Provider>
`;
